File: backend/schemas/live_capture.py
from typing import Any, Optional

from pydantic import BaseModel, Field


class LiveCaptureRequest(BaseModel):
  interface: Optional[str] = Field(default=None, max_length=64)
  bpf_filter: str = Field(default="ip or ip6", max_length=256)
  packet_count: int = Field(default=50, ge=1, le=5000)
  timeout_seconds: int = Field(default=30, ge=1, le=300)
  min_packets_before_predict: int = Field(default=2, ge=1, le=100)
  send_to_backend: bool = True
  session_id: Optional[str] = Field(default=None, max_length=100)


class LiveCaptureSummary(BaseModel):
  attack_counts: dict[str, int] = Field(default_factory=dict)
  threat_levels: dict[str, int] = Field(default_factory=dict)
  top_threat: Optional[dict[str, Any]] = None
  benign_flows: int = 0
  malicious_flows: int = 0


class LiveCaptureDetection(BaseModel):
  flow_id: str
  source_ip: Optional[str] = None
  destination_ip: Optional[str] = None
  source_port: Optional[int] = None
  destination_port: Optional[int] = None
  protocol: str
  total_packets: int
  prediction: dict[str, Any]
  detected_at: str


class LiveCaptureResponse(BaseModel):
  success: bool = True
  message: str
  session_id: str
  status: str
  interface: Optional[str] = None
  bpf_filter: str
  packets_processed: int
  predictions_generated: int
  threats_detected: int
  summary: LiveCaptureSummary
  detections: list[LiveCaptureDetection] = Field(default_factory=list)
  backend_delivery: Optional[dict[str, Any]] = None
  completed_at: str
  job_id: Optional[str] = None


class LiveCaptureCapabilitiesResponse(BaseModel):
  success: bool = True
  message: str
  enabled: bool
  default_interface: Optional[str] = None
  default_bpf_filter: str
  default_packet_count: int
  default_timeout_seconds: int
  send_to_backend: bool
  models_loaded: list[str] = Field(default_factory=list)
  notes: list[str] = Field(default_factory=list)

File: frontend/src/types/liveCapture.ts
export interface LiveCaptureRequest {
  interface?: string | null;
  bpf_filter?: string;
  packet_count?: number;
  timeout_seconds?: number;
  min_packets_before_predict?: number;
  send_to_backend?: boolean;
  session_id?: string | null;
}

export interface LiveCaptureSummary {
  attack_counts?: Record<string, number>;
  threat_levels?: Record<string, number>;
  top_threat?: {
    attack_label: string;
    count: number;
  };
  benign_flows?: number;
  malicious_flows?: number;
}

export interface LiveCaptureDetection {
  flow_id: string;
  source_ip?: string | null;
  destination_ip?: string | null;
  source_port?: number | null;
  destination_port?: number | null;
  protocol: string;
  total_packets: number;
  prediction: {
    attack_label: string;
    confidence: number;
    is_attack: boolean;
    threat_level: string;
  };
  detected_at: string;
}

export interface LiveCaptureResponse {
  success: boolean;
  message: string;
  session_id: string;
  status: string;
  interface?: string | null;
  bpf_filter: string;
  packets_processed: number;
  predictions_generated: number;
  threats_detected: number;
  summary: LiveCaptureSummary;
  detections: LiveCaptureDetection[];
  backend_delivery?: Record<string, unknown> | null;
  completed_at: string;
  job_id?: string | null;
}

export interface LiveCaptureCapabilitiesResponse {
  success: boolean;
  message: string;
  enabled: boolean;
  default_interface?: string | null;
  default_bpf_filter: string;
  default_packet_count: number;
  default_timeout_seconds: number;
  send_to_backend: boolean;
  models_loaded: string[];
  notes: string[];
}
